### PR TITLE
Wip Updatehub Support

### PR DESCRIPTION
**Pull request by chtavares592**

### CODE REVIEW

File: subsys/updatehub/Kconfig [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+    select NET_TCP
+    select HTTP
+    select HTTP_CLIENT
+    select NET_APP_NEED_IPV4
+    select NET_APP_SETTINGS
+    select DNS_RESOLVER
+    select JSON_LIBRARY
+    select TINYCRYPT
+    select TINYCRYPT_SHA256
+    default"n"
```
Comment (otavio) [inline subsys/updatehub/Kconfig]:
```
default "n"
```

File: subsys/updatehub/updatehub.c [hunk 2]
```diff
@@ -0,0 +1,852 @@
...
+
+#define HTTP_NETWORK_TIMEOUT  K_SECONDS(CONFIG_HTTP_CLIENT_NETWORK_TIMEOUT)
+#define TCP_RECV_BUFFER_SIZE 2048
+#define MAX_URL_SIZE 300
+#define SEND_BUFFER_SIZE 430
+
+#define UPDATEHUB_SHA256_SIZE 32
+
+#define UPDATEHUB_SERVER "10.5.3.14"    /* api.updatehub.io */
+#define UPDATEHUB_PORT 8000             /* 80 */
```
Comment (otavio) [inline subsys/updatehub/updatehub.c]: This should be removed now and set to the real value.

File: subsys/updatehub/updatehub.c [hunk 3]
```diff
@@ -0,0 +1,852 @@
...
+	if (confirmed_image == false) {
+		k_delayed_work_submit(&updatehub_context.work,
+				      K_SECONDS(1));
+	}
+	if (confirmed_image == true) {
+		k_delayed_work_submit(&updatehub_context.work,
+				      K_MINUTES(CONFIG_UPDATEHUB_POLL_INTERVAL));
+	}
+
+}
```
Comment (otavio) [inline subsys/updatehub/updatehub.c]:
```
static void work_submit_queue(struct k_delayed_work *work)
{
	if (!confirmed_image == false) {
                k_delayed_work_submit(&updatehub_context.work,
				      K_SECONDS(1));
                return;
	}

	k_delayed_work_submit(&updatehub_context.work,
			      K_MINUTES(CONFIG_UPDATEHUB_POLL_INTERVAL));
}
```

File: subsys/updatehub/updatehub.c [hunk 4]
```diff
@@ -0,0 +1,852 @@
...
+
+static enum updatehub_response _install_update(struct updatehub_context *ctx)
+{
+	int ret = 0;
+	enum updatehub_response response;
+
+	tc_sha256_init(&sha256sum);
+
+	ret = init_flash();
```
Comment (otavio) [inline subsys/updatehub/updatehub.c]: No need to use a function as it is called only here.

File: include/updatehub.h [hunk 5]
```diff
@@ -0,0 +1,75 @@
+/*
+ * Copyright (c) 2018 O.S.Systems
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/** @file
+ *
+ *  @brief UpdateHub Firmware Over-the-Air library for Zephyr.
```
Comment (otavio) [inline include/updatehub.h]: `... for Zephyr Project`

File: subsys/updatehub/updatehub.c [hunk 6]
```diff
@@ -0,0 +1,912 @@
...
+#define MAX_PAYLOAD_SIZE 430
+#define SHA256SUM_STRING_SIZE 65
+#define MAX_FRAG_SIZE 129
+#define COAP_MAX_TRY_GET_PKT 3
+
+#define API_UPDATEHUB "10.5.3.67" /* TODO: replace by the right URL */
+#define API_DOWNLOAD_UPDATEHUB ""
+
+struct updatehub_coap {
+	struct net_context *ctx;
```
Comment (otavio) [inline subsys/updatehub/updatehub.c]: Use `struct net_context *net_ctx;`

File: subsys/updatehub/updatehub.c [hunk 7]
```diff
@@ -0,0 +1,912 @@
...
+
+static int metadata_hash_get(char *metadata)
+{
+	int ret = 0, i;
+	char buffer[2];
+	unsigned char metadata_hash[TC_SHA256_DIGEST_SIZE];
+
+	ret = tc_sha256_init(&sha256sum);
+	if (ret < 1) {
+		SYS_LOG_ERR("Could not start sha256sum");
```
Comment (otavio) [inline subsys/updatehub/updatehub.c]: Move this to an `error` label and jump to it using `goto`.

File: subsys/updatehub/updatehub.c [hunk 8]
```diff
@@ -0,0 +1,912 @@
...
+		return ret;
+	}
+
+	ret = tc_sha256_update(&sha256sum, metadata, strlen(metadata));
+	if (ret < 1) {
+		SYS_LOG_ERR("Could not update sha256sum");
```
Comment (otavio) [inline subsys/updatehub/updatehub.c]: `goto`.

File: subsys/updatehub/updatehub.c [hunk 9]
```diff
@@ -0,0 +1,912 @@
...
+		return ret;
+	}
+
+	ret = tc_sha256_final(metadata_hash, &sha256sum);
+	if (ret < 1) {
+		SYS_LOG_ERR("Could not finish sha256sum");
```
Comment (otavio) [inline subsys/updatehub/updatehub.c]: `goto`.

File: subsys/updatehub/updatehub.c [hunk 10]
```diff
@@ -0,0 +1,912 @@
...
+		return ret;
+	}
+
+	for (i = 0; i < TC_SHA256_DIGEST_SIZE; i++) {
+		snprintk(buffer, TC_SHA256_DIGEST_SIZE, "%02x", metadata_hash[i]);
+		strcat(&probe_storage.package_uid[i], buffer);
+	}
+
+	return 1;
```
Comment (otavio) [inline subsys/updatehub/updatehub.c]: Success is `0`, isn't it?

File: subsys/updatehub/updatehub.c [hunk 11]
```diff
@@ -0,0 +1,912 @@
...
+	/* TODO: Is need the build_num? */
+	snprintk(firmware_version,
+		 BOOT_IMG_VER_STRLEN_MAX,
+		 "%d.%d.%d.%d",
+		 image_header.h.v1.sem_ver.major,
+		 image_header.h.v1.sem_ver.minor,
+		 image_header.h.v1.sem_ver.revision,
+		 image_header.h.v1.sem_ver.build_num);
+
+	return 0;
```
Comment (otavio) [inline subsys/updatehub/updatehub.c]: `return ret`;

File: subsys/updatehub/updatehub.c [hunk 12]
```diff
@@ -0,0 +1,936 @@
...
+#include <flash.h>
+#include <dfu/mcuboot.h>
+#include <dfu/flash_img.h>
+#include <tinycrypt/sha256.h>
+
+#include <updatehub.h>
+#include "updatehub_priv.h"
+#include "device_identity.h"
+
+#define UPDATEHUB_NETWORK_TIMEOUT K_SECONDS(10)
```
Comment (otavio) [inline subsys/updatehub/updatehub.c]: Use `NETWORK_TIMEOUT` as this is local to this file.

File: subsys/updatehub/updatehub.c [hunk 13]
```diff
@@ -0,0 +1,936 @@
...
+#define MAX_PATH_SIZE 255
+#define MAX_PAYLOAD_SIZE 430
+#define SHA256SUM_STRING_SIZE 65
+#define MAX_FRAG_SIZE 129
+#define COAP_MAX_TRY_GET_PKT 3
```
Comment (otavio) [inline subsys/updatehub/updatehub.c]: `COAP_MAX_RETRY`

File: subsys/updatehub/updatehub.c [hunk 14]
```diff
@@ -0,0 +1,936 @@
...
+	struct coap_block_context block;
+	enum updatehub_response code_status;
+	u8_t uri_path[MAX_PATH_SIZE];
+	u8_t payload[MAX_PAYLOAD_SIZE];
+	int downloaded_size;
+};
+
+struct updatehub_context {
+	struct k_delayed_work work;
+};
```
Comment (otavio) [inline subsys/updatehub/updatehub.c]: Couldn't we get rid of this script and just pass `struct k_delayed_work` instance around?

File: subsys/updatehub/updatehub.c [hunk 15]
```diff
@@ -0,0 +1,936 @@
...
+	int ret = -1;
+
+	for (i = 0; i < metadata_some_boards->supported_hardware_len; i++) {
+		if (strcmp(metadata_some_boards->supported_hardware[i],
+			   CONFIG_BOARD) == 0) {
+			ret = 0;
+			break;
+		}
+	}
+	return ret;
```
Comment (otavio) [inline subsys/updatehub/updatehub.c]: Newline before the `return`

File: subsys/updatehub/updatehub.c [hunk 16]
```diff
@@ -0,0 +1,936 @@
...
+}
+
+static void install_update_cb(struct net_context *ctx,
+			      struct net_pkt *pkt,
+			      int status, void *user)
+{
+	int ret = -1, i;
```
Comment (otavio) [inline subsys/updatehub/updatehub.c]: Split this in two lines. You must adopt a style for declaring it.

File: subsys/updatehub/updatehub.c [hunk 17]
```diff
@@ -0,0 +1,936 @@
...
+			goto end;
+		}
+	}
+
+	coap.code_status = UPDATEHUB_OK;
+
+end:
+	net_pkt_unref(pkt);
+	k_sem_give(&updatehub_sem);
+	return;
```
Comment (otavio) [inline subsys/updatehub/updatehub.c]: This `return` is not required.

File: subsys/updatehub/updatehub.c [hunk 18]
```diff
@@ -0,0 +1,936 @@
...
+
+	SYS_LOG_DBG("Metadata received:\n%s", metadata_buffer);
+
+	coap.code_status = UPDATEHUB_OK;
+
+end:
+	net_pkt_unref(pkt);
+	k_sem_give(&updatehub_sem);
+	return;
+}
```
Comment (otavio) [inline subsys/updatehub/updatehub.c]: The probe and install callback are mostly identical. See if you can rework the code to avoid the duplication.

Comment (chtavares592) [reply subsys/updatehub/updatehub.c]: I already tried to separate the parts that seem equal, but when I do that I can't get the flag of metadata so I receive the metadata empty.

**Pull request closed without merging**